do.py

```diff
@@ -0,0 +1,1174 @@
+#! /usr/bin/env python3
+
+'''
+Developer CLI: building (meson), tests, benchmark, etc.
+
+This file contains tasks definitions for doit (https://pydoit.org).
+And also a CLI interface using click (https://click.palletsprojects.com).
+
+The CLI is ideal for project contributors while,
+doit interface is better suited for authring the development tasks.
+
+REQUIREMENTS:
+--------------
+- PyPI packages: doit==0.35.0, click, rich-click==1.3.0
+
+# USAGE:
+
+## 1 - click API
+
+Commands can added using default Click API. i.e.
+
+```
+@cli.command()
+@click.argument('extra_argv', nargs=-1)
+@click.pass_obj
+def python(ctx_obj, extra_argv):
+    """Start a Python shell with PYTHONPATH set"""
+```
+
+## 2 - class based Click command definition
+
+`CliGroup` provides an alternative class based API to create Click commands.
+
+Just use the `cls_cmd` decorator. And define a `run()` method
+
+```
+@cli.cls_cmd('test')
+class Test():
+    """Run tests"""
+
+    @classmethod
+    def run(cls):
+        print('Running tests...')
+```
+
+- Command may make use a Click.Group context defining a `ctx` class attribute
+- Command options are also define as class attributes
+
+```
+@cli.cls_cmd('test')
+class Test():
+    """Run tests"""
+    ctx = CONTEXT
+
+    verbose = Option(['--verbose', '-v'], default=False, is_flag=True, help="more verbosity")
+
+    @classmethod
+    def run(cls, **kwargs): # kwargs contains options from class and CONTEXT
+        print('Running tests...')
+```
+
+## 3 - This class based interface can be run as a doit task simply by subclassing from Task
+
+- Extra doit task metadata can be defined as class attribute TASK_META.
+- `run()` method will be used as python-action by task
+
+```
+@cli.cls_cmd('test')
+class Test(Task):   # Task base class, doit will create a task
+    """Run tests"""
+    ctx = CONTEXT
+
+    TASK_META = {
+        'task_dep': ['build'],
+    }
+
+    @classmethod
+    def run(cls, **kwargs):
+        pass
+```
+
+## 4 - doit tasks with cmd-action "shell" or dynamic metadata
+
+Define method `task_meta()` instead of `run()`:
+
+```
+@cli.cls_cmd('refguide-check')
+class RefguideCheck(Task):
+    @classmethod
+    def task_meta(cls, **kwargs):
+        return {
+```
+
+
+
+# TODO:
+First milestone replace dev.py
+
+- [ ] doc: it would make sense to expose MAKE targets as sub-commands or options
+
+cleanup + doit:
+- [ ] move out non-scipy code
+- [ ] doit reporter when running under click
+
+BUG:
+- [ ] python dev.py -t scipy.optimize.tests.test_minimize_constrained.py::TestTrustRegionConstr::test_args
+      unknown marker "slow". seems conftest is not being loaded
```

Comment:
I thought this was solved with latest pytest. They had a regression on 7.0 and 7.0.1 should fix it.

Reply:
It still fails for me on pytest 7.1.1
Note the command line above reproduces an error on existing `dev.py`, not on the code in this PR.

Reply:
Failed for me as well using `pytest 7.7.1`

```shell
pytest.PytestUnknownMarkWarning: Unknown pytest.mark.slow 
```

Reply:
@sayantikabanik feel free to look into this. I am busy with other stuff...

Reply:
:facepalm: The problem is the example is using python module notation (separated by dot), but also includes `.py`.
A pytest bug is responsible for the misleading error message...

WRONG:
```
python dev.py -t scipy.optimize.tests.test_minimize_constrained.py::TestTrustRegionConstr::test_args
```

CORRECT:
```
python dev.py -t scipy.optimize.tests.test_minimize_constrained::TestTrustRegionConstr::test_args
python dev.py -t scipy/optimize/tests/test_minimize_constrained.py::TestTrustRegionConstr::test_args
```